tensorflow/contrib/s3/s3_file_system.cc

```diff
@@ -0,0 +1,546 @@
+    };
+    Aws::InitAPI(options);
+  }
+  ~S3FileSystem() {
+    Aws::SDKOptions options;
+    options.loggingOptions.logLevel = Aws::Utils::Logging::LogLevel::Info;
+    Aws::ShutdownAPI(options);
+  }
+  Status NewRandomAccessFile(
+      const string& fname, std::unique_ptr<RandomAccessFile>* result) override {
+    string bucket, object;
+    TF_RETURN_IF_ERROR(ParseS3Path(fname, false, &bucket, &object));
+    result->reset(new S3RandomAccessFile(bucket, object));
+    return Status::OK();
+  }
+  Status NewWritableFile(const string& fname,
+                         std::unique_ptr<WritableFile>* result) override {
+    string bucket, object;
+    TF_RETURN_IF_ERROR(ParseS3Path(fname, false, &bucket, &object));
+    result->reset(new S3WritableFile(bucket, object));
+    return Status::OK();
+  }
+
+  Status NewAppendableFile(const string& fname,
+                           std::unique_ptr<WritableFile>* result) override {
+    std::unique_ptr<RandomAccessFile> reader;
+    TF_RETURN_IF_ERROR(NewRandomAccessFile(fname, &reader));
+    std::unique_ptr<char[]> buffer(new char[S3ReadAppendableFileBufferSize]);
+    Status status;
+    uint64 offset = 0;
+    StringPiece read_chunk;
+
+    string bucket, object;
+    TF_RETURN_IF_ERROR(ParseS3Path(fname, false, &bucket, &object));
+    result->reset(new S3WritableFile(bucket, object));
+
+    while (true) {
+      status = reader->Read(offset, S3ReadAppendableFileBufferSize, &read_chunk,
+                            buffer.get());
+      if (status.ok()) {
+        (*result)->Append(read_chunk);
+        offset += S3ReadAppendableFileBufferSize;
+      } else if (status.code() == error::OUT_OF_RANGE) {
+        (*result)->Append(read_chunk);
+        break;
+      } else {
+        (*result).reset();
+        return status;
+      }
+    }
+
+    return Status::OK();
+  }
+
+  Status NewReadOnlyMemoryRegionFromFile(
+      const string& fname,
+      std::unique_ptr<ReadOnlyMemoryRegion>* result) override {
+    uint64 size;
+    TF_RETURN_IF_ERROR(GetFileSize(fname, &size));
+    std::unique_ptr<char[]> data(new char[size]);
+
+    std::unique_ptr<RandomAccessFile> file;
+    TF_RETURN_IF_ERROR(NewRandomAccessFile(fname, &file));
+
+    StringPiece piece;
+    TF_RETURN_IF_ERROR(file->Read(0, size, &piece, data.get()));
+
+    result->reset(new S3ReadOnlyMemoryRegion(std::move(data), size));
+    return Status::OK();
+  }
+
+  Status FileExists(const string& fname) override {
+    FileStatistics stats;
+    TF_RETURN_IF_ERROR(this->Stat(fname, &stats));
+    return Status::OK();
+  }
+
+  Status GetChildren(const string& dir, std::vector<string>* result) override {
+    string bucket, prefix;
+    TF_RETURN_IF_ERROR(ParseS3Path(dir, false, &bucket, &prefix));
+
+    if (prefix.back() != '/') {
+      prefix.push_back('/');
+    }
+
+    Aws::S3::S3Client s3Client;
+    Aws::S3::Model::ListObjectsV2Request listObjectsV2Request;
```

Comment:
I noticed that you used `listObjectV2`. In some compatible S3 implementations from 3rd-party, the new V2 interface is not supported. I think it is better to provide an option for using `listObject` (V1). With the V1 inferface, TensorFlow-S3 can work with more 3rd-party S3 implementations.

I tried to remove all `V2`, and change `ContinuationToken` into `Marker`. It seems working well.

Thanks!